extensions/amp-ad/0.1/amp-ad-api-handler.js

```diff
@@ -0,0 +1,205 @@
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {removeElement} from '../../../src/dom';
+import {loadPromise} from '../../../src/event-helper';
+import {
+  listenFor,
+  listenForOnce,
+  postMessage,
+} from '../../../src/iframe-helper';
+import {parseUrl} from '../../../src/url';
+import {IntersectionObserver} from '../../../src/intersection-observer';
+import {viewerFor} from '../../../src/viewer';
+import {user} from '../../../src/log';
+
+export class AmpAdApiHandler {
+
+  /**
+   * @param {!BaseElement} baseInstance
+   * @param {!Element} element
+   * @param {function()=} opt_noContentCallback
+   */
+  constructor(baseInstance, element, opt_noContentCallback) {
+    /** @private {!BaseElement} */
+    this.baseInstance_ = baseInstance;
+
+    /** @privat {!Element} */
+    this.element_ = element;
+
+    /** @private {?Element} iframe instance */
+    this.iframe_ = null;
+
+    /** @private {?IntersectionObserver} */
+    this.intersectionObserver_ = null;
+
+    /** @private {boolean} */
+    this.is3p_ = false;
+
+    /** @param {?function()} opt_noContentHandler */
+    this.noContentCallback_ = opt_noContentCallback;
+
+    /** @private {!Array<!Function>} functions to unregister listeners */
+    this.unlisteners_ = [];
+
+    /** @private @const */
+    this.viewer_ = viewerFor(this.baseInstance_.getWin());
+  }
+
+  /**
+   * @param {!Element} iframe
+   * @param {boolean} is3p
+   * @return {!Promise} awaiting load event for ad frame
+   */
+  startUp(iframe, is3p) {
+    user.assert(
+      !this.iframe, 'multiple invocations of startup without destroy!');
+    this.iframe_ = iframe;
+    this.is3p_ = is3p;
+    this.iframe_.setAttribute('scrolling', 'no');
+    this.baseInstance_.applyFillContent(this.iframe_);
+    this.intersectionObserver_ =
+        new IntersectionObserver(this.baseInstance_, this.iframe_, is3p);
+    // Triggered by context.noContentAvailable() inside the ad iframe.
+    listenForOnce(this.iframe_, 'no-content', () => {
+      if (this.noContentCallback_) {
+        this.noContentCallback_();
+      } else {
+        user.info('no content callback was specified');
+      }
+    }, this.is3p_);
+    // Triggered by context.reportRenderedEntityIdentifier(…) inside the ad
+    // iframe.
+    listenForOnce(this.iframe_, 'entity-id', info => {
+      this.element_.creativeId = info.id;
+    }, this.is3p_);
+    this.unlisteners_.push(listenFor(this.iframe_, 'embed-size', data => {
+      let newHeight, newWidth;
+      if (data.width !== undefined) {
+        newWidth = Math.max(this.element_./*OK*/offsetWidth +
+            data.width - this.iframe_./*OK*/offsetWidth, data.width);
+        this.iframe_.width = newWidth;
```

Comment:
@bobcassels  I'm wondering why we changed

```
this.iframe_.width = data.width;
```

to 

```
 this.iframe_.width = newWidth;
```

See the [code before the change](https://github.com/ampproject/amphtml/pull/3534/files#diff-03f204ef56e45e6c843922d48e709848L275). 

Was it intentional to cope with something? 

cc @zhouyx